### PR TITLE
Add control command basic tests

### PR DESCRIPTION
**Pull request by yontyon**
In this PR we introduce 2 basic Apollo tests for the Wedge command.
1. Test that the system stops as a response to a wedge command when everything goes well
2. Test that the system stops as a response to a wedge command even when one replica is getting updated by the state transfer mechanism.

This PR shows that the basic functionality of the wedge command is working

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -2382,8 +2382,11 @@ void ReplicaImp::onTransferringCompleteImp(SeqNum newStateCheckpoint) {
 void ReplicaImp::onSeqNumIsSuperStable(SeqNum newSuperStableSeqNum) {
   auto seq_num_to_stop_at = controlStateManager_->getCheckpointToStopAt();
   if (seq_num_to_stop_at.has_value() && seq_num_to_stop_at.value() == newSuperStableSeqNum) {
-    if (getRequestsHandler()->getControlHandlers())
+    LOG_INFO(GL, "control state manager marked seqNum: " << newSuperStableSeqNum);
+    if (getRequestsHandler()->getControlHandlers()) {
+      metric_on_call_back_of_super_stable_cp_.Get().Set(1);
       getRequestsHandler()->getControlHandlers()->onSuperStableCheckpoint();
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.cpp:2388]: The more I read `SuperStableCheckpoint`, the less I like it. This doesn't have to be done in this PR, as it will have ripple effects, but maybe we can change it to something like `NofNStableCheckpoint`. Then we can rename the callback to be `onNofNCheckpointsReceived`

I'm open to other naming ideas as well. If everyone disagrees, we can also keep it as is :)

File: tests/apollo/test_skvbc_control_commands.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    @with_trio
+    @with_bft_network(start_replica_cmd,  selected_configs=lambda n, f, c: n == 4 and c == 0)
+    async def test_wedge_command(self, bft_network):
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        client = bft_network.random_client()
+        rep = await client.write(bytearray(), special_flags=0x9)
+        await skvbc.write_known_kv()
+        for replica_id in range(bft_network.config.n):
```
Comment (andrewjstone) [inline tests/apollo/test_skvbc_control_commands.py:57]: I assume this check works because the primary sends no-ops to fill the windows, right?

File: tests/apollo/test_skvbc_control_commands.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            "-t", os.environ.get('STORAGE_TYPE')]
+
+
+class SkvbcControlCommandsTest(unittest.TestCase):
+
+    # __test__ = False  # so that PyTest ignores this test scenario
+
+    @with_trio
+    @with_bft_network(start_replica_cmd,  selected_configs=lambda n, f, c: n == 4 and c == 0)
+    async def test_wedge_command(self, bft_network):
```
Comment (andrewjstone) [inline tests/apollo/test_skvbc_control_commands.py:51]: Can you please add a doc comment explaining what this function does? See our other apollo tests for guidance on what they should look like. This is especially important so that the final try clause is documented as explicitly failing because the system shouldn't accept more commands.

File: util/pyclient/bft_msgs.py [hunk 4]
```diff
@@ -53,13 +53,14 @@ def __init__(self, message):
 ReplyHeader = namedtuple('ReplyHeader', ['span_context_size', 'primary_id',
     'req_seq_num', 'length', 'rsi_length'])
 
-def pack_request(client_id, req_seq_num, read_only, timeout_milli, cid, msg, pre_process=False, span_context=b''):
+def pack_request(client_id, req_seq_num, read_only, timeout_milli, cid, msg, pre_process=False, span_context=b'', special_flags=None):
```
Comment (andrewjstone) [inline util/pyclient/bft_msgs.py:56]: This API addition is way too low-level. It requires the caller to know the hex value of the special flags. While that could be replaced with some global constant, it's better to instead not expose the flags, which are a low level detail of the protocol, and can be changed without needing to change the `pack_request` API.

Instead you should expose the behavior you want. In this case I think it would be better to continue with the pattern here and take another boolean that defaults to false : `wedge=false` or `stop_consensus=false`, and then sent the proper flag value in the function. This will also make the code at the call sites be more legible such that readers can understand the behavior without having to figure out what special flags are.

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 5]
```diff
@@ -41,6 +41,17 @@ int InternalCommandsHandler::execute(uint16_t clientId,
                                      uint32_t &outActualReplySize,
                                      uint32_t &outActualReplicaSpecificInfoSize,
                                      concordUtils::SpanWrapper &span) {
+  bool wedgeCommand = flags & 0x9;
```
Comment (andrewjstone) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:44]: `0x9` doesn't really work as a flag, as it uses two bits: bit 0, and bit 3. If someone added a flag of `0x8`, then `0x8` and `0x1` (read-only) combined would match a wedge command and lead to incorrect behavior. You should just use `0x4` or `0x8` instead.

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 6]
```diff
@@ -41,6 +41,17 @@ int InternalCommandsHandler::execute(uint16_t clientId,
...
+  if (wedgeCommand) {
+    LOG_INFO(m_logger, "A wedge command has been called seqNum: " << sequenceNum);
```
Comment (andrewjstone) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:46]: Please use the KVLOG wrapper for sequenceNum so this log message looks the same as others. I realize this is just test output, but it's good to get into the habit.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.cpp:2388]: let's leach it for a future PR :)

File: tests/apollo/test_skvbc_control_commands.py (hunk 2 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py:57]: It was a mistake.
Fixed

File: util/pyclient/bft_msgs.py (hunk 4 above)
Comment (yontyon) [reply util/pyclient/bft_msgs.py:56]: fixed

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp (hunk 5 above)
Comment (yontyon) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:44]: I have removed the flag completely (as wedge command should be a write command)

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp (hunk 6 above)
Comment (yontyon) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:46]: done

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 7]
```diff
@@ -2382,8 +2382,11 @@ void ReplicaImp::onTransferringCompleteImp(SeqNum newStateCheckpoint) {
 void ReplicaImp::onSeqNumIsSuperStable(SeqNum newSuperStableSeqNum) {
   auto seq_num_to_stop_at = controlStateManager_->getCheckpointToStopAt();
   if (seq_num_to_stop_at.has_value() && seq_num_to_stop_at.value() == newSuperStableSeqNum) {
-    if (getRequestsHandler()->getControlHandlers())
+    LOG_INFO(GL, "control state manager marked seqNum: " << newSuperStableSeqNum);
```
[the remark this answers is not in the document]
Comment (yontyon) [reply bftengine/src/bftengine/ReplicaImp.cpp:2385]: done

**Comment by yontyon** [comment]
Guys, if you have ideas for more testing, please let me know :)

**Comment by teoparvanov** [comment]
> 
> 
> Guys, if you have ideas for more tests, please let me know :)

As an edge case, I wonder what would happen if you send a request while a minority (<f) replicas are doing state transfer, so they ignore the wedge command.

File: tests/apollo/test_skvbc_control_commands.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        Sends a wedge command and check that the system stops from processing new requests.
+        Note that in this test we assume no failures and synchronized network.
+        The test does the following:
+        1. A client sends a wedge command
+        2. The client verify that the system reached to a super stable checkpoint
+        3. The client tries to initiate a new write bft command and fails
+    """
+
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 4 and c == 0)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:56]: If you want to run this only for a single configuration, let's make it `lambda n, f, c: n == 7`, as it's closer to the real environment.
Actually, why don't you want to run the test for all supported configs?

File: tests/apollo/test_skvbc_control_commands.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
+# Concord
+#
+# Copyright (c) 2019 VMware, Inc. All Rights Reserved.
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:3]: nit - 2020

File: tests/apollo/CMakeLists.txt [hunk 10]
```diff
@@ -76,5 +80,6 @@ foreach(STORAGE_TYPE ${STORAGE_TYPES})
             "env ${APOLLO_TEST_ENV} STORAGE_TYPE=${STORAGE_TYPE} python3 -m unittest test_skvbc_ro_replica ${TEST_OUTPUT}"
             WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
   endif()
+
```
Comment (teoparvanov) [inline tests/apollo/CMakeLists.txt:83]: nit - remove empty line

File: tests/apollo/test_skvbc_control_commands.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 4 and c == 0)
+    async def test_wedge_command(self, bft_network):
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        client = bft_network.random_client()
+
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=0)
+
+        last_bid = skvbc.parse_reply(await client.read(skvbc.get_last_block_req()))
+        await client.write(skvbc.write_req([], [], block_id=last_bid, wedge_command=True))
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:65]: no need to set the `block_id` field - it's only used if your read-set is non-empty

File: tests/apollo/test_skvbc_control_commands.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        The test does the following:
+        1. Start all replicas but 1
+        2. A client sends a wedge command
+        3. Validate that all started replicas reached to the next next checkpoint
+        4. Start the late replica
+        5. Validate that the late replica completed the state transfer
+        6. Validate that all replicas stopped at the super stable checkpoint and that new commands are not being processed
+    """
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 4 and c == 0)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:89]: same here - why not run for all configs? or at least for `n==7` ?

File: tests/apollo/test_skvbc_control_commands.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    async def test_wedge_command_with_state_transfer(self, bft_network):
+        initial_prim = 0
+        late_replica = bft_network.random_set_of_replicas(1, {initial_prim})
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:92]: `late_replicas`

File: tests/apollo/test_skvbc_control_commands.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        on_time_replicas = bft_network.all_replicas(without=late_replica)
+        bft_network.start_replicas(on_time_replicas)
+
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=0)
+
+        client = bft_network.random_client()
+        last_bid = skvbc.parse_reply(await client.read(skvbc.get_last_block_req()))
+        await client.write(skvbc.write_req([], [], block_id=last_bid, wedge_command=True))
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:102]: same here - no need to set `block_id`

File: tests/apollo/test_skvbc_control_commands.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+                        if checkpoint_after == checkpoint_before + 2:
+                            break
+
+        bft_network.start_replicas(late_replica)
+
+        await bft_network.wait_for_state_transfer_to_start()
+        for r in late_replica:
+            await bft_network.wait_for_state_transfer_to_stop(initial_prim,
+                                                              r,
+                                                              stop_on_stable_seq_num=False)
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:118]: isn't `stop_on_stable_seq_num=False` by default ?

File: tests/apollo/test_skvbc_control_commands.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    viewChangeTimeoutMilli = "10000"
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    return [path,
+            "-k", KEY_FILE_PREFIX,
+            "-i", str(replica_id),
+            "-s", statusTimerMilli,
+            "-v", viewChangeTimeoutMilli,
+            "-p" if os.environ.get('BUILD_ROCKSDB_STORAGE', "").lower()
+                    in set(["true", "on"])
+            else "",
```
Comment (teoparvanov) [inline tests/apollo/test_skvbc_control_commands.py:41]: formatting: the `else` should be under the if

**Comment by teoparvanov** [comment]
> 
> 
> Guys, if you have ideas for more tests, please let me know :)

What happens if the primary is down and at this point you send a wedge command? Will the wedge command be ignored? Will it be processed after the view change completes? I think this needs to be tested (in a separate PR, of course). WDYT?

**Comment by andrewjstone** [comment]
> > Guys, if you have ideas for more tests, please let me know :)
> 
> What happens if the primary is down and at this point you send a wedge command? Will the wedge command be ignored? Will it be processed after the view change completes? I think this needs to be tested (in a separate PR, of course). WDYT?

One thing to note about this code is that its happy path only. We specifically aren't building a fault tolerant upgrade at this point. We could probably make it so that the wedge would be processed after the view change completes, and should, I'm just not sure it's in scope for this change. The immediate, short term goal is if anything goes wrong to do a manual recovery. 

We definitely need to have a chat about that...

With all the above being said, I actually don't think its that hard to make sure the wedge is processed after view change completes and we'll need to consider that in the future. It should also be noted that the `superStableCheckpoint` i.e. N out of N checkpoints received will change to normal N-F in the future as well, when we can tolerate failures. This is just an intermediate step.

File: tests/apollo/test_skvbc_control_commands.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+            with trio.fail_after(seconds=30):
+                while True:
+                    with trio.move_on_after(seconds=1):
+                        checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=replica_id)
+                        if checkpoint_after == checkpoint_before + 2:
+                            break
+
+        await self.validate_stop_on_super_stable_checkpoint(bft_network, skvbc)
+
+    """
```
Comment (andrewjstone) [inline tests/apollo/test_skvbc_control_commands.py:77]: The doc comment should be placed right under the function definition, not before it.

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp [hunk 18]
```diff
@@ -111,6 +111,10 @@ bool InternalCommandsHandler::executeWriteCommand(uint32_t requestSize,
                << " PRE_PROCESS_FLAG=" << ((flags & MsgFlag::PRE_PROCESS_FLAG) != 0 ? "true" : "false")
                << " HAS_PRE_PROCESSED_FLAG=" << ((flags & MsgFlag::HAS_PRE_PROCESSED_FLAG) != 0 ? "true" : "false"));
 
+  if (writeReq->header.type == WEDGE) {
+    LOG_INFO(m_logger, "A wedge command has been called in seqNum: " << KVLOG(sequenceNum));
```
Comment (andrewjstone) [inline tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp:115]: Nit: No need for "in seqNum". KVLOG will print the output as `sequenceNum: <ActualValue>`.

File: tests/apollo/test_skvbc_control_commands.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+            "-s", statusTimerMilli,
+            "-v", viewChangeTimeoutMilli,
+            "-p" if os.environ.get('BUILD_ROCKSDB_STORAGE', "").lower()
+                    in set(["true", "on"])
+            else "",
+            "-t", os.environ.get('STORAGE_TYPE')]
+
+
+class SkvbcControlCommandsTest(unittest.TestCase):
+    """
```
Comment (andrewjstone) [inline tests/apollo/test_skvbc_control_commands.py:46]: The doc comment should be under the function definition, not before it.

File: tests/apollo/test_skvbc_control_commands.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        for replica_id in on_time_replicas:
+            with trio.fail_after(seconds=30):
+                while True:
+                    with trio.move_on_after(seconds=1):
+                        checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=replica_id)
+                        if checkpoint_after == checkpoint_before + 2:
+                            break
+
+        bft_network.start_replicas(late_replica)
+
```
Comment (andrewjstone) [inline tests/apollo/test_skvbc_control_commands.py:113]: I think it might be useful to add an additional check here that tries to send another request even though we haven't seen a super stable checkpoint yet. At this point each replica should stop accepting requests since it has reached the sequence number to stop at. The waiting for NofN is just to provide the control state callback to start the upgrade.

Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py:113]: I think that in this case, the late replica will initiate a view change (eventually).
@teoparvanov, what do you think?

**Comment by yontyon** [comment]
> > Guys, if you have ideas for more tests, please let me know :)
> 
> What happens if the primary is down and at this point you send a wedge command? Will the wedge command be ignored? Will it be processed after the view change completes? I think this needs to be tested (in a separate PR, of course). WDYT?

As @andrewjstone explained, we are mainly interested in the happy case only (i.e. all replicas are up and the network is synchronized).
Yet, following the design, if we managed to execute the command, eventually, all replicas (even thus that were updated by the state transfer) should stop at this super stable checkpoint.
I also think that would be a good test case for future PR.

File: tests/apollo/test_skvbc_control_commands.py (hunk 8 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: Fixed. Run for all configurations

File: tests/apollo/test_skvbc_control_commands.py (hunk 9 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed

File: tests/apollo/CMakeLists.txt (hunk 10 above)
Comment (yontyon) [reply tests/apollo/CMakeLists.txt]: fixed

File: tests/apollo/test_skvbc_control_commands.py (hunk 11 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed

File: tests/apollo/test_skvbc_control_commands.py (hunk 12 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: Fixed. Run for all configurations

File: tests/apollo/test_skvbc_control_commands.py (hunk 13 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed

File: tests/apollo/test_skvbc_control_commands.py (hunk 14 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed

File: tests/apollo/test_skvbc_control_commands.py (hunk 15 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: set to true

File: tests/apollo/test_skvbc_control_commands.py (hunk 16 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed

File: tests/apollo/test_skvbc_control_commands.py (hunk 17 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed

File: tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp (hunk 18 above)
Comment (yontyon) [reply tests/simpleKVBC/TesterReplica/internalCommandsHandler.cpp]: fixed

File: tests/apollo/test_skvbc_control_commands.py (hunk 19 above)
Comment (yontyon) [reply tests/apollo/test_skvbc_control_commands.py]: fixed
